### PR TITLE
Websocket update increasing exchange coverage and bug fixes

### PR DESCRIPTION
**Pull request by shazbert**
# Description

Light fixes websocket functionality

Fixes # (issue)

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Running the binary and checking output.

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Please also consider improving test coverage whilst working on a certain package

- [x] Test A
- [ ] Test B

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool 
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Travis with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by thrasher-** [comment]
No worries @Daanikus!

File: exchanges/poloniex/poloniex_websocket.go [hunk 1]
```diff
@@ -88,7 +87,18 @@ func (p *Poloniex) WsSubscribe() error {
...
+func (p *Poloniex) WsReadData() (exchange.WebsocketResponse, error) {
+	_, resp, err := p.WebsocketConn.ReadMessage()
+	if err != nil {
+		return exchange.WebsocketResponse{}, err
+	}
+
+	p.Websocket.TrafficAlert <- struct{}{}
+	return exchange.WebsocketResponse{Raw: resp}, nil
+}
+
```
[the remark this answers is not in the document]
Comment (Daanikus) [reply exchanges/poloniex/poloniex_websocket.go]: @CodeLingoBot review

**Comment by Daanikus** [comment]
@CodeLingoBot review

**Comment by shazbert** [comment]
@Daanikus Just to let you know Codelingo is still running.

File: exchanges/zb/zb_websocket.go [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+
+			z.Websocket.DataHandler <- exchange.WebsocketOrderbookUpdate{
+				Pair:     cPair,
+				Asset:    "SPOT",
+				Exchange: z.GetName(),
+			}
+
+		case common.StringContains(result.Channel, "trades"):
+			trades := struct {
+				Data []struct {
```
Comment (gloriousCode) [inline exchanges/zb/zb_websocket.go]: Could you move inline structs to zb_types.go?

File: exchanges/gateio/gateio_websocket.go [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+			Params: []interface{}{c.Pair().String()},
+		}
+
+		err = g.WebsocketConn.WriteJSON(trade)
+		if err != nil {
+			return err
+		}
+
+		depth := WebsocketRequest{
+			ID:     1337,
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/gateio/gateio_websocket.go]: In the api docs they are needed, only for different websocket sessions which we don't need just yet.

File: exchanges/gateio/gateio.go [hunk 4]
```diff
@@ -8,6 +8,8 @@ import (
 	"strings"
 	"time"
 
+	"github.com/gorilla/websocket"
+
```
Comment (thrasher-) [inline exchanges/gateio/gateio.go]: Remove newline

File: exchanges/gateio/gateio_websocket.go [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/thrasher-/gocryptotrader/exchanges/orderbook"
+
```
Comment (thrasher-) [inline exchanges/gateio/gateio_websocket.go]: Remove various newlines

File: exchanges/gateio/gateio_websocket.go [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+				err = common.JSONDecode(result.Params[1], &data)
+				if err != nil {
+					g.Websocket.DataHandler <- err
+					continue
+				}
+
+				var asks, bids []orderbook.Item
+				askData, ok := data["asks"]
+				if ok {
```
Comment (thrasher-) [inline exchanges/gateio/gateio_websocket.go]: throw an err if we can't access the value of the asks and bids keys

File: exchanges/huobihadax/huobihadax.go [hunk 7]
```diff
@@ -9,6 +9,8 @@ import (
 	"strconv"
 	"time"
 
+	"github.com/gorilla/websocket"
+
```
Comment (thrasher-) [inline exchanges/huobihadax/huobihadax.go]: remove newline

File: exchanges/bitmex/bitmex_websocket.go [hunk 8]
```diff
@@ -184,7 +176,7 @@ func (b *Bitmex) wsHandleIncomingData() {
 			}
 
 			quickCapture := make(map[string]interface{})
-			err := common.JSONDecode(resp.Raw, &quickCapture)
+			err = common.JSONDecode(resp.Raw, &quickCapture)
 			if err != nil {
 				log.Error(err)
```
Comment (thrasher-) [inline exchanges/bitmex/bitmex_websocket.go]: Send err to DataHandler chan and continue

File: exchanges/btcc/btcc_websocket.go [hunk 9]
```diff
@@ -85,45 +84,47 @@ func (b *BTCC) WsConnect() error {
...
 	for {
 		select {
 		case <-b.Websocket.ShutdownC:
 			return
 
-		case resp := <-b.Websocket.Intercomm:
+		default:
+			resp, err := b.WsReadData()
+			if err != nil {
+				b.Websocket.DataHandler <- err
```
Comment (thrasher-) [inline exchanges/btcc/btcc_websocket.go]: return

File: exchanges/coinbasepro/coinbasepro_websocket.go [hunk 10]
```diff
@@ -110,37 +120,19 @@ func (c *CoinbasePro) WsReadData() {
 			return
 
 		default:
-			_, resp, err := c.WebsocketConn.ReadMessage()
+			resp, err := c.WsReadData()
 			if err != nil {
 				c.Websocket.DataHandler <- err
```
Comment (thrasher-) [inline exchanges/coinbasepro/coinbasepro_websocket.go]: return

File: exchanges/coinut/coinut_websocket.go [hunk 11]
```diff
@@ -46,31 +57,13 @@ func (c *COINUT) WsReadData() {
 			return
 
 		default:
-			_, resp, err := c.WebsocketConn.ReadMessage()
+			resp, err := c.WsReadData()
 			if err != nil {
 				c.Websocket.DataHandler <- err
```
Comment (thrasher-) [inline exchanges/coinut/coinut_websocket.go]: return

File: exchanges/gemini/gemini.go [hunk 12]
```diff
@@ -8,6 +8,8 @@ import (
 	"strings"
 	"time"
 
+	"github.com/gorilla/websocket"
+
```
Comment (thrasher-) [inline exchanges/gemini/gemini.go]: remove newline

File: exchanges/gemini/gemini_websocket.go [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+		case <-g.Websocket.ShutdownC:
+			return
+
+		case resp := <-comms:
+			switch resp.FeedType {
+			case geminiWsEvent:
+
+			case geminiWsMarketData:
+				var result Response
+				common.JSONDecode(resp.Raw, &result)
```
Comment (thrasher-) [inline exchanges/gemini/gemini_websocket.go]: check result of JSONDecode using err var

File: exchanges/huobi/huobi_websocket.go [hunk 14]
```diff
@@ -79,41 +102,14 @@ func (h *HUOBI) WsReadData() {
...
-
-	for {
-		select {
-		case <-h.Websocket.ShutdownC:
-		case resp := <-h.Websocket.Intercomm:
 			var init WsResponse
-			err := common.JSONDecode(resp.Raw, &init)
+			err = common.JSONDecode(resp.Raw, &init)
 			if err != nil {
 				log.Error(err)
```
Comment (thrasher-) [inline exchanges/huobi/huobi_websocket.go]: Send to DataHandler channel and continue

File: exchanges/huobihadax/huobihadax_websocket.go [hunk 15]
```diff
@@ -0,0 +1,341 @@
...
+			resp, err := h.WsReadData()
+			if err != nil {
+				h.Websocket.DataHandler <- err
+				return
+			}
+
+			var init WsResponse
+			err = common.JSONDecode(resp.Raw, &init)
+			if err != nil {
+				log.Error(err)
```
Comment (thrasher-) [inline exchanges/huobihadax/huobihadax_websocket.go]: Send err to DataHandler channel and continue for this one and the ones below

File: exchanges/poloniex/poloniex_websocket.go [hunk 16]
```diff
@@ -151,16 +130,29 @@ func checkSubscriptionSuccess(data []interface{}) bool {
...
 	for {
 		select {
 		case <-p.Websocket.ShutdownC:
 			return
 
-		case resp := <-p.Websocket.Intercomm:
+		default:
+			resp, err := p.WsReadData()
+			if err != nil {
+				p.Websocket.DataHandler <- err
```
Comment (thrasher-) [inline exchanges/poloniex/poloniex_websocket.go]: return

File: exchanges/poloniex/poloniex_websocket.go [hunk 17]
```diff
@@ -151,16 +130,29 @@ func checkSubscriptionSuccess(data []interface{}) bool {
...
+			}
+
 			var result interface{}
-			err := common.JSONDecode(resp.Raw, &result)
+			err = common.JSONDecode(resp.Raw, &result)
 			if err != nil {
```
Comment (thrasher-) [inline exchanges/poloniex/poloniex_websocket.go]: Send to DataHandler error channel and continue

File: exchanges/okex/okex_websocket.go [hunk 18]
```diff
@@ -188,17 +166,30 @@ func (o *OKEX) wsPingHandler() {
...
 	for {
 		select {
 		case <-o.Websocket.ShutdownC:
 			return
 
-		case resp := <-o.Websocket.Intercomm:
+		default:
+			resp, err := o.WsReadData()
+			if err != nil {
+				o.Websocket.DataHandler <- err
```
Comment (thrasher-) [inline exchanges/okex/okex_websocket.go]: add return

**Pull request merged**